File: Tools/1 - Common/MoCommon/Geom/SIntPadding.cs
using MO.Common.IO;
using MO.Common.Lang;

namespace MO.Common.Geom
{
   //============================================================
   // <T>整数边界。</T>
   //
   // @history MAOCY 121025
   //============================================================
   public class SIntPadding
   {
      // 左边界
      public int Left;

      // 上边界
      public int Top;

      // 右边界
      public int Right;

      // 下边界
      public int Bottom;

      //============================================================
      // <T>构造整数边界。</T>
      //============================================================
      public SIntPadding(int left = 0, int top = 0, int right = 0, int bottom = 0) {
         Left = left;
         Top = top;
         Right = right;
         Bottom = bottom;
      }

      //============================================================
      // <T>判定是否为空。</T>
      //
      // @return 是否为空
      //============================================================
      public bool IsEmpty() {
         return (0 == Left) && (0 == Top) && (0 == Right) && (0 == Bottom);
      }

      //============================================================
      // <T>接收对象内容。</T>
      //
      // @param padding 对象内容
      //============================================================
      public void Assign(SIntPadding padding) {
         if(padding != null) {
            Left = padding.Left;
            Top = padding.Top;
            Right = padding.Right;
            Bottom = padding.Bottom;
         }
      }
      
      //============================================================
      // <T>重置数据。</T>
      //============================================================
      public void Reset() {
         Left = 0;
         Top = 0;
         Right = 0;
         Bottom = 0;
      }

      //============================================================
      // <T>解析字符串内容。</T>
      //
      // @param value 字符串内容
      //============================================================
      public bool Parse(string value) {
         if (!RString.IsEmpty(value)) {
            string[] items = value.Split(',');
            if(4 == items.Length) {
               Left = RInt.Parse(items[0]);
               Top = RInt.Parse(items[1]);
               Right = RInt.Parse(items[2]);
               Bottom = RInt.Parse(items[3]);
               return true;
            }
         }
         return false;
      }

      //============================================================
      // <T>序列化内容到输出流。</T>
      //
      // @param output 输出流
      //============================================================
      public void Serialize8(IOutput output) {
         output.WriteUint8((byte)Left);
         output.WriteUint8((byte)Top);
         output.WriteUint8((byte)Right);
         output.WriteUint8((byte)Bottom);
      }

      //============================================================
      // <T>序列化内容到输出流。</T>
      //
      // @param output 输出流
      //============================================================
      public void Serialize16(IOutput output) {
         output.WriteUint16((ushort)Left);
         output.WriteUint16((ushort)Top);
         output.WriteUint16((ushort)Right);
         output.WriteUint16((ushort)Bottom);
      }

      //============================================================
      // <T>序列化内容到输出流。</T>
      //
      // @param output 输出流
      //============================================================
      public void Serialize(IOutput output) {
         output.WriteInt32(Left);
         output.WriteInt32(Top);
         output.WriteInt32(Right);
         output.WriteInt32(Bottom);
      }

      //============================================================
      // <T>获得字符串内容。</T>
      //
      // @return 字符串内容
      //============================================================
      public override string ToString() {
         return Left + "," + Top + "," + Right + "," + Bottom;
      }
   }
}

File: Source/Cross/Common/MoMath/SIntPadding.cpp
#include "MoMtGeom.h"
#include "MoCmLanguage.h"

MO_NAMESPACE_BEGIN

//============================================================
// <T>序列化数据内容到输出流。</T>
// 
// @param pOutput 输出流
// @return 处理结果
//============================================================
TResult SIntPadding::Serialize8(IDataOutput* pOutput){
   MO_CHECK(pOutput, return ENull);
   pOutput->WriteUint8((TUint8)left);
   pOutput->WriteUint8((TUint8)top);
   pOutput->WriteUint8((TUint8)right);
   pOutput->WriteUint8((TUint8)bottom);
   return ESuccess;
}

//============================================================
// <T>从输入流反序列化数据内容。</T>
// 
// @param pInput 输入流
// @return 处理结果
//============================================================
TResult SIntPadding::Unserialize8(IDataInput* pInput){
   MO_CHECK(pInput, return ENull);
   left = pInput->ReadUint8();
   top = pInput->ReadUint8();
   right = pInput->ReadUint8();
   bottom = pInput->ReadUint8();
   return ESuccess;
}


//============================================================
// <T>序列化数据内容到输出流。</T>
// 
// @param pOutput 输出流
// @return 处理结果
//============================================================
TResult SIntPadding::Serialize16(IDataOutput* pOutput){
   MO_CHECK(pOutput, return ENull);
   pOutput->WriteUint16((TUint16)left);
   pOutput->WriteUint16((TUint16)top);
   pOutput->WriteUint16((TUint16)right);
   pOutput->WriteUint16((TUint16)bottom);
   return ESuccess;
}

//============================================================
// <T>从输入流反序列化数据内容。</T>
// 
// @param pInput 输入流
// @return 处理结果
//============================================================
TResult SIntPadding::Unserialize16(IDataInput* pInput){
   MO_CHECK(pInput, return ENull);
   left = pInput->ReadUint16();
   top = pInput->ReadUint16();
   right = pInput->ReadUint16();
   bottom = pInput->ReadUint16();
   return ESuccess;
}

//============================================================
// <T>序列化数据内容到输出流。</T>
// 
// @param pOutput 输出流
// @return 处理结果
//============================================================
TResult SIntPadding::Serialize(IDataOutput* pOutput){
   MO_CHECK(pOutput, return ENull);
   pOutput->WriteInt32((TInt32)left);
   pOutput->WriteInt32((TInt32)top);
   pOutput->WriteInt32((TInt32)right);
   pOutput->WriteInt32((TInt32)bottom);
   return ESuccess;
}

//============================================================
// <T>从输入流反序列化数据内容。</T>
// 
// @param pInput 输入流
// @return 处理结果
//============================================================
TResult SIntPadding::Unserialize(IDataInput* pInput){
   MO_CHECK(pInput, return ENull);
   left = pInput->ReadInt32();
   top = pInput->ReadInt32();
   right = pInput->ReadInt32();
   bottom = pInput->ReadInt32();
   return ESuccess;
}

MO_NAMESPACE_END
